examples/diffdope_ros/scripts/server.py

```diff
@@ -0,0 +1,342 @@
+
+    def __get_camera_intrinsic_params(self, camera_info_topic_name):
+        camera_info = None
+
+        def camera_info_callback(data):
+            nonlocal camera_info
+            camera_info = data
+
+        rospy.Subscriber(camera_info_topic_name, CameraInfo, camera_info_callback)
+
+        while camera_info is None and not rospy.is_shutdown():
+            rospy.sleep(0.1)
+
+        fx = camera_info.K[0]  # Focal length in x-axis
+        fy = camera_info.K[4]  # Focal length in y-axis
+        cx = camera_info.K[2]  # Principal point in x-axis
+        cy = camera_info.K[5]  # Principal point in y-axis
+
+        image_height = camera_info.height
+        image_width = camera_info.width
+
+        Camera = namedtuple("Camera", "fx fy cx cy image_height image_width")
+        camera_intrinsic_parameters = Camera(fx, fy, cx, cy, image_height, image_width)
+        return camera_intrinsic_parameters
+
+    def __refine_object(self, goal):
+        refined_pose = self.__compute_refined_pose(goal.target_object)
+        result = RefineObjectResult()
+        result.refined_estimation = refined_pose
+        self._action_refine_object_server.set_succeeded(result)
+
+    def __refine_all(self, goal):
+        result = RefineAllResult()
+        for target_object in goal.target_objects:
+            refined_pose = self.__compute_refined_pose(target_object)
+            result.refined_estimations.append(refined_pose)
+        self._action_refine_all_server.set_succeeded(result)
+
+    def __compute_refined_pose(self, target_object):
+        initial_pose_estimate = target_object.pose.pose
+
+        scene = self.__generate_scene(
+            target_object.rgb_frame,
+            target_object.depth_frame,
+            initial_pose_estimate.position,
+        )
+        object3d = self.__generate_3d_object(target_object)
+
+        ddope = dd.DiffDope(cfg=self.cfg, scene=scene, object3d=object3d)
+        ddope.run_optimization()
+
+        if self.cfg.save_video:
+            self.__save_video(ddope)
+
+        refined_pose = self.__create_pose_stamped(ddope.get_pose())
+        target_object = TargetObject()
+        target_object.name = target_object.name
+        target_object.pose = refined_pose
+
+        return target_object
+
+    def __save_video(self, ddope):
+        rospy.loginfo("***You can disable video outputting from the config file***")
+        video_dir = os.path.expanduser(self.cfg.saved_videos_path)
+        self.__create_dir_if_not_exists(video_dir)
+        next_video_id = self.__find_next_available_video_id(video_dir)
+        video_path = os.path.join(video_dir, f"video_{next_video_id}.mp4")
+        ddope.make_animation(output_file_path=video_path)
+        rospy.loginfo(f"Video saved at {video_path}")
+
+    def __convert_opengl_pose_to_opencv(self, transform):
+        transform[0, 3] /= 10
+        transform[1, 3] /= 10
+        transform[2, 3] /= 10
```

Comment:
One small issue that is currently hard-coded is that I observed the output of diff-dope was off by a fixed error.  Here, I divide the output by 10 to correct it. I understand this is not ideal and ugly, but I am not sure where this error comes from. This is strange because it seems that Diff-DOPE expects the initial position estimate in millimetres, yet the output seems more close to metres? The 3D model is scaled to match the scale of your provided example by 10, but I don't understand how this would affect the output.

Reply:
Yeah I had some issues in there, and these are somehow are to pin point, thank you for this, maybe we should make this a hyper params, as the 3d model will have an impact on the representation. I have a vague recollection that I needed this, but did it just to match my older code.